File: HW/HW12/Main.py
class Edge:
    def __init__(self, from_, to, cost):
        self.from_ = from_
        self.to = to
        self.cost = cost

class DisjointSet:
    def __init__(self, size):
        self.parent = [i for i in range(size + 1)]
        self.rank = [0] * (size + 1)
        
    def find(self, x):
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]
    
    def union(self, x, y):
        root_x = self.find(x)
        root_y = self.find(y)
        
        if root_x == root_y:
            return
        
        if self.rank[root_x] < self.rank[root_y]:
            self.parent[root_x] = root_y
        elif self.rank[root_x] > self.rank[root_y]:
            self.parent[root_y] = root_x
        else:
            self.parent[root_y] = root_x
            self.rank[root_x] += 1

class Solution:
    def minCostToSupplyWater(self, n, wells, pipes):
        e_list = []

        for p in pipes:
            house1, house2, cost = p
            e_list.append(Edge(house1, house2, cost))

        for i, well_cost in enumerate(wells):
            e_list.append(Edge(0, i + 1, well_cost))

        e_list.sort(key=lambda edge: edge.cost)

        d = DisjointSet(n)
        total = 0
        for e in e_list:
            from_ = d.find(e.from_)
            to = d.find(e.to)

            if from_ != to:
                total += e.cost
                d.union(from_, to)

        return total

File: HW/HW12/Main.java
public class hw12 {

	class Edge {
		int from, to, cost;

		Edge(int from, int to, int cost) {
			this.from = from; this.to = to; this.cost = cost;
		}
	}
	
	class DisjointSet {
		private int[] parent, rank;
		
		public DisjointSet(int size) {
			parent = new int[size+1]; rank = new int[size+1];
			for (int i = 0; i <= size; i++) {
				parent[i] = i;
				rank[i] = 0;
			}
		}
		
		public int find(int x) {
			if (parent[x] != x) {
				parent[x] = find(parent[x]);
			}
			return parent[x];
		}
		
		public void union(int x, int y) {
			int rootX = find(x), rootY = find(y);
			
			if (rootX == rootY) {
				return;
			}
			
			if (rank[rootX] < rank[rootY]) {
				parent[rootX] = rootY;
			} else if (rank[rootX] > rank[rootY]) {
				parent[rootY] = rootX;
			} else {
				parent[rootY] = rootX;
				rank[rootX]++;
			}
		}
	}

	public int minCostToSupplyWater(int n, int[] wells, int[][] pipes) {

		List<Edge> eList = new ArrayList<>();

		for (int[] p : pipes) {
			int house1 = p[0], house2 = p[1], cost = p[2];
			eList.add(new Edge(house1, house2, cost));
		}

		for (int i = 0; i < wells.length; i++) {
			eList.add(new Edge(0, i + 1, wells[i]));
		}

		Collections.sort(eList, Comparator.comparingInt(edge -> edge.cost));
		
		DisjointSet d = new DisjointSet(n);
		
		int total = 0;
		for (Edge e: eList) {
			int from = d.find(e.from), to = d.find(e.to);
			
			if (from != to) {
				total += e.cost;
				d.union(from, to);
			}
		}

		return total;
	}
}
